tests/integration/functions/types_test.ts

```diff
@@ -0,0 +1,162 @@
+import { assert } from "../../../src/dev_deps.ts";
+import { FunctionRuntimeType } from "../../../src/functions/types.ts";
+import { DefineFunction, Schema, SlackAPI } from "../../../src/mod.ts";
+import { CanBe } from "../../../src/test_utils.ts";
+
+Deno.test("FunctionRuntimeType should abe able to provide a usable type of a DefineFunction return object", () => {
+  const testFunctionDefinition = DefineFunction({
+    callback_id: "test_function",
+    title: "Test function",
+    source_file: "functions/test_function.ts",
+    input_parameters: {
+      properties: {
+        bool: {
+          type: Schema.types.boolean,
+        },
+        int: {
+          type: Schema.types.integer,
+        },
+        num: {
+          type: Schema.types.number,
+        },
+        string: {
+          type: Schema.types.string,
+        },
+        arr: {
+          type: Schema.types.array,
+          items: {
+            type: Schema.types.boolean,
+          },
+        },
+        obj: {
+          type: Schema.types.object,
+          properties: {
+            bool: {
+              type: Schema.types.boolean,
+            },
+          },
+          required: ["bool"],
+        },
+      },
+      required: ["bool", "int", "num", "string", "arr", "obj"],
+    },
+    output_parameters: {
+      properties: {
+        bool: {
+          type: Schema.types.boolean,
+        },
+        int: {
+          type: Schema.types.integer,
+        },
+        num: {
+          type: Schema.types.number,
+        },
+        string: {
+          type: Schema.types.string,
+        },
+        arr: {
+          type: Schema.types.array,
+          items: {
+            type: Schema.types.boolean,
+          },
+        },
+        obj: {
+          type: Schema.types.object,
+          properties: {
+            bool: {
+              type: Schema.types.boolean,
+            },
+          },
+          required: ["bool"],
+        },
+      },
+      required: ["bool", "int", "num", "string", "arr", "obj"],
+    },
+  });
+
+  type Actual = FunctionRuntimeType<typeof testFunctionDefinition>;
+
+  const expected: Actual = {
+    outputs: {
+      bool: true,
+      int: 0,
+      num: 0.1,
+      string: "",
+      arr: [true],
+      obj: {
+        bool: true,
+      },
+    },
+    args: {
+      inputs: {
+        bool: true,
+        int: 0,
+        num: 0.1,
+        string: "",
+        arr: [true],
+        obj: {
+          bool: true,
+        },
+      },
+      enterprise_id: "",
+      env: {},
+      team_id: "",
+      token: "",
+      client: SlackAPI("", {}),
+      event: {
+        event_ts: "",
+        function: {
+          id: "",
+          callback_id: "",
+          title: "",
+        },
+        function_execution_id: "",
+        type: "function_executed",
+      },
+    },
+  };
+
+  assert<CanBe<typeof expected, Actual>>(true);
```

Comment:
I think this assertion will always be true, since we manually assign the `Actual` type to the `expected` variable.

You're on the right track, though, for how to test this.. hmm.. maybe one thing we can try instead is to actually implement the `testFunctionDefinition` runtime logic / handler, and then fake 'executing' it to retrieve the output of the function, then compare the type of the function handler output to the type returned by the `FunctionRuntimeType` type?

For an example of this, check out the [output parameter integration tests](https://github.com/slackapi/deno-slack-sdk/blob/main/tests/integration/functions/runtime_context/output_parameters_test.ts):

- [here](https://github.com/slackapi/deno-slack-sdk/blob/main/tests/integration/functions/runtime_context/output_parameters_test.ts#L25-L32) we implement a function handler that meets the function definition in the test. You could do a similar thing in this test (and move a majority of what the `expected` variable you have in this test sets to inside the function handler). The nice thing about the various `*SlackFunctionHandler` SDK types is that they force the dev writing the handler to return the required output parameters, as per their function definition.
- [here](https://github.com/slackapi/deno-slack-sdk/blob/main/tests/integration/functions/runtime_context/output_parameters_test.ts#L33-L34) we use some of our SDK integration testing utilities to create a function runtime 'context' object (the thing that has all the extras like token, env vars, team_id, etc.) useful for testing, and feed it to the handler we implemented on line 25.
- with the above two steps, we can then retrieve the output from the function! Once you have the actual runtime output from the function, you can use the exact kind of `assert` you use here on it to make sure the types are compatible.

Hit me up if you wanna pair on this, happy to help!

Reply:
Thanks for the information 💯  I've implemented this and it seems to work